src/v/kafka/server/replicated_partition.cc

```diff
@@ -92,32 +92,41 @@ ss::future<model::record_batch_reader> replicated_partition::make_reader(
         ss::lw_shared_ptr<const storage::offset_translator_state> _translator;
     };
     auto rdr = co_await _partition->make_reader(cfg, deadline);
-    co_return model::make_record_batch_reader<reader>(
-      std::move(rdr).release(), _translator);
+    co_return storage::translating_reader(
+      model::make_record_batch_reader<reader>(
+        std::move(rdr).release(), _translator),
+      _translator);
 }
 
 ss::future<std::vector<cluster::rm_stm::tx_range>>
 replicated_partition::aborted_transactions(
-  model::offset base, model::offset last) {
-    model::offset local_kafka_start_offset = _translator->from_log_offset(
-      _partition->start_offset());
-    if (base < local_kafka_start_offset) {
-        // TODO: get offset translation information for the offsets range
-        // that we have read and use it to query
-        // _partition->aborted_transactions.
-        co_return std::vector<cluster::rm_stm::tx_range>{};
-    }
+  model::offset base,
+  model::offset last,
+  ss::lw_shared_ptr<const storage::offset_translator_state> ot_state) {
+    vassert(ot_state, "ntp {}: offset translator state must be present", ntp());
 
-    auto source = co_await _partition->aborted_transactions(
-      _translator->to_log_offset(base), _translator->to_log_offset(last));
+    // Note: here we expect that local _partition contains aborted transaction
+    // ids for both local and remote offset ranges. This is true as long as
+    // rm_stm state has not been reset (for example when there is a partition
+    // transfer or when a stale replica recovers its log from beyond the log
+    // eviction point). See https://github.com/vectorizedio/redpanda/issues/3001
+
+    auto base_rp = ot_state->to_log_offset(base);
+    auto last_rp = ot_state->to_log_offset(last);
+    auto source = co_await _partition->aborted_transactions(base_rp, last_rp);
 
     std::vector<cluster::rm_stm::tx_range> target;
     target.reserve(source.size());
     for (const auto& range : source) {
+        // Trim aborted ranges to the range bounds of read offsets so that we
+        // are guaranteed to have offset translation data for these offsets.
+        auto trimmed_first = std::max(base_rp, range.first);
```

Comment:
We can't trim it because we pass it to the client and the clients may depend on it. I understand this PR is a best effort until we support #3001 so it might be ok but we should not affect users who don't use shadow indexes. Is it possible to check if we deal with local or cloud data and to trim only if it's cloud?